sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -482,6 +482,15 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
 
   // LongConverter
   private[this] def castToLong(from: DataType): Any => Any = from match {
+    case StringType if ansiEnabled =>
```

Comment:
I'd like to see something like
```
case StringType if ansiEnabled =>
  val result = new LongWrapper()
  buildCast[UTF8String](_, s => {
    s.toLongExact(result)
    result.value
  }
case StringType =>
  val result = new LongWrapper()
  buildCast[UTF8String](_, s => if (s.toLong(result)) result.value else null)
```
and in codegen
```
val method = if (ansi) "toIntExact" else "toInt"
code"""
  UTF8String.IntWrapper $wrapper = new UTF8String.IntWrapper();
  if ($c.$method($wrapper)) {
    $evPrim = $wrapper.value;
  } else {
    $evNull = true;
  }
  $wrapper = null;
```

Reply:
For long, won't it be good to use `LongWrapper` ?

Reply:
`LongWrapper` is only needed in `toLong` because we have to return 2 values: the long and a boolean flag. This is not true anymore in `toLongExact`.

Reply:
Oh, Thanks for clarification. :)

I guess here LongWrapper is required .
`UTF8String.IntWrapper $wrapper = new UTF8String.IntWrapper();`

For Int,Short and Byte, format should be like this only or previous changes are fine for them?

Reply:
Let's create `toByteExact`, `toShortExact`, `toIntExact` and `toLongExact` all together, and use similar approach to implement the cast for them.

Reply:
The implementation of `toLongExact` will be almost similar to `toLong`. Won't it be duplication of code? 
What about this implementation ->

```
  public long toLongExact() {
    LongWrapper result = new LongWrapper();
    if (toLong(result)) {
      return result.value;
    }
    throw new NumberFormatException("invalid input syntax for type numeric: " + this);
  }
```

Here, i can use the `toLong` code for checking whether string contains valid value or not.

If we are not using the wrapper then implementation will be almost similar, as i will be throwing exception whenever a invalid value is there instead of returning `false` in `toLongExact`.

Reply:
I'm fine with it. In the future, if we only have the ANSI behavior, we can unify the code.

Reply:
Ok. I will create `toByteExact`, `toShortExact`, `toIntExact` and `toLongExact` using the existing format.